src/shared/openssl-util.c

```diff
@@ -137,4 +223,159 @@ int string_hashsum(
+        int nid = NID_undef;
+        int r;
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+        _cleanup_free_ char *curve_name = NULL;
+        size_t len = 0;
+        if (EVP_PKEY_get_group_name(pkey, NULL, 0, &len) != 1 || len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine PKEY group name length");
+        len += 1;
+        curve_name = malloc(len);
+        if (!curve_name)
+                return log_oom();
+
+        if (!EVP_PKEY_get_group_name(pkey, curve_name, len, &len))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get PKEY group name");
+        nid = OBJ_sn2nid(curve_name);
+#else
+        EC_KEY *ec_key = EVP_PKEY_get0_EC_KEY(pkey);
+        if (!ec_key)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "PKEY doesn't have EC_KEY associated");
+
+        if (EC_KEY_check_key(ec_key) != 1)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "EC_KEY associated with PKEY is not valid");
+
+        nid = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec_key));
+#endif
+
+        r = pkey_generate_ec_key(nid, &pkey_new);
+        if (r < 0)
+                return log_error_errno(r, "Failed to generate ec key: %m");
+
+        r = pkey_ecdh_derive_shared_secret(pkey_new, pkey, NULL, ret_decrypted_key_size);
+        if (r < 0 || *ret_decrypted_key_size == 0)
+                return log_error_errno(r, "Failed to determine derived shared secret size: %m");
+
+        *ret_decrypted_key = malloc(*ret_decrypted_key_size);
+        if (!*ret_decrypted_key)
+                return log_oom();
+
+        r = pkey_ecdh_derive_shared_secret(pkey_new, pkey, *ret_decrypted_key, ret_decrypted_key_size);
+        if (r < 0)
+                return log_error_errno(r, "Failed to derive shared secret: %m");
+
+        *ret_savedata_size = i2d_PUBKEY(pkey_new, NULL);
+        if (*ret_savedata_size == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine encoded public key size");
+
+        *ret_savedata = malloc(*ret_savedata_size);
+        if (!*ret_savedata)
+                return log_oom();
+
+        /* i2d_PUBKEY function has a side effect that makes *pp point to end of the allocated buffer */
+        uint8_t *buffer = *ret_savedata;
+        *ret_savedata_size = i2d_PUBKEY(pkey_new, &buffer);
+        if (*ret_savedata_size == 0)
+                return  log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get encoded public key");
+
+        return 0;
+
+}
+
+static int pkey_generate_volume_key_rsa(
+                EVP_PKEY *pkey,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                void **ret_savedata,
+                size_t *ret_savedata_size) {
+
+        assert(pkey);
+        assert(ret_decrypted_key);
+        assert(ret_savedata);
+        assert(ret_savedata_size);
+
+        int r = rsa_pkey_to_suitable_key_size(pkey, ret_decrypted_key_size);
```

Comment:
so `int r` is our special, favourite variable to store return codes of functions in. Because of that it gets this special privilege, to always be defined uptop, and not "lazily" where it is needed. (doesn't matter though)